File: app/src/main/java/com/example/cameraaccess/viewmodel/CreateSiteBlockViewModel.kt
package com.example.cameraaccess.viewmodel

import android.app.Application
import android.util.Log
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import com.example.cameraaccess.data.model.ApiCache
import com.example.cameraaccess.data.model.SiteBlockCreateRequest
import com.example.cameraaccess.data.repositories.SiteBlockRepository
import com.example.cameraaccess.ui.createsiteblock.CreateSiteBlockUiState
import com.google.gson.Gson

class CreateSiteBlockViewModel(app: Application) : AndroidViewModel(app) {

    private val repo = SiteBlockRepository(app)
    private val cache = ApiCache(app)

    val uiState = mutableStateOf(CreateSiteBlockUiState())

    init {
        val data = repo.getCachedSiteBlock()
        if (data == null) {
            uiState.value = uiState.value.copy(
                error = "No cached data. Go back and press New Scan again."
            )
        } else {
            uiState.value = uiState.value.copy(data = data)
        }
    }

    // -------------------------
    // Form setters (UI calls these)
    // -------------------------
    fun setSelectedSiteIndex(i: Int) {
        uiState.value = uiState.value.copy(
            selectedSiteIndex = i,
            selectedBlockIndex = -1 // reset block when site changes
        )
    }

    fun setSelectedBlockIndex(i: Int) {
        uiState.value = uiState.value.copy(selectedBlockIndex = i)
    }

    fun setSelectedScanTypeIndex(i: Int) {
        uiState.value = uiState.value.copy(selectedScanTypeIndex = i)
    }

    fun setRowWidth(v: String) {
        uiState.value = uiState.value.copy(rowWidth = v)
    }

    fun setRowHeight(v: String) {
        uiState.value = uiState.value.copy(rowHeight = v)
    }

    fun setBayLength(v: String) {
        uiState.value = uiState.value.copy(bayLength = v)
    }

    fun setAddLens(v: Boolean) {
        uiState.value = uiState.value.copy(
            addLens = v,
            multiplier = if (!v) "" else uiState.value.multiplier
        )
    }

    fun setMultiplier(v: String) {
        uiState.value = uiState.value.copy(multiplier = v)
    }

    // -------------------------
    // Submit (called by UI)
    // -------------------------
    fun submit() {
        val state = uiState.value
        val data = state.data ?: run {
            uiState.value = uiState.value.copy(error = "Missing cached site data.")
            return
        }

        val sites = data.sites
        val scanTypes = data.scan_type

        // Validate selections
        if (state.selectedSiteIndex !in sites.indices) return
        val blocks = sites[state.selectedSiteIndex].blocks
        if (state.selectedBlockIndex !in blocks.indices) return
        if (state.selectedScanTypeIndex !in scanTypes.indices) return

        // Validate numbers
        val rw = state.rowWidth.toDoubleOrNull() ?: return
        val rh = state.rowHeight.toDoubleOrNull() ?: return
        val bl = state.bayLength.toDoubleOrNull() ?: return
        val addedMultiplier =
            if (state.addLens) state.multiplier.toDoubleOrNull() ?: 1.0 else 1.0

        val body = SiteBlockCreateRequest(
            blockId = blocks[state.selectedBlockIndex].block_id,
            scanTypeId = scanTypes[state.selectedScanTypeIndex].id,
            rowWidth = rw,
            rowHeight = rh,
            bayLength = bl,
            addedMultiplier = addedMultiplier
        )
        
        uiState.value = uiState.value.copy(isLoading = true, error = null)

        try {
            // Convert the request object to a JSON string
            val json = Gson().toJson(body)
            Log.d("SiteBlockDebug", "submit() saving to cache: body=$json")

            // Save the JSON string to the local cache
            cache.saveSubmitSiteBlock(json)
            cache.saveUseMainRearCameraForScan(state.addLens)

            // Trigger navigation
            uiState.value = uiState.value.copy(
                isLoading = false,
                navigateNext = true
            )
        } catch (e: Exception) {
            Log.e("SiteBlockDebug", "Failed to save to cache", e)
            uiState.value = uiState.value.copy(
                isLoading = false,
                error = "Failed to save settings: ${e.message}"
            )
        }
    }

    fun consumeNavigation() {
        uiState.value = uiState.value.copy(navigateNext = false)
    }
}

File: TauResearch/ViewModels/CreateSiteBlockViewModel.swift
// CreateSiteBlockViewModel.swift
// Mirrors CreateSiteBlockViewModel.kt + CreateSiteBlockUiState.

import Foundation
import Observation

struct CreateSiteBlockUiState {
    var data: SiteBlockResponse?
    var isLoading: Bool = false
    var error: String?
    var navigateNext: Bool = false

    var selectedSiteIndex: Int = -1
    var selectedBlockIndex: Int = -1
    var selectedScanTypeIndex: Int = -1

    var rowWidth: String = ""
    var rowHeight: String = ""
    var bayLength: String = ""

    var addLens: Bool = false
    var multiplier: String = ""
}

@Observable
@MainActor
final class CreateSiteBlockViewModel {

    private let repo = SiteBlockRepository()
    var state = CreateSiteBlockUiState()

    init() {
        if let cached = repo.getCachedSiteBlock() {
            state.data = cached
        } else {
            state.error = "No cached data. Go back and press New Scan again."
        }
    }

    // MARK: - Setters

    func setSelectedSiteIndex(_ i: Int) {
        state.selectedSiteIndex = i
        state.selectedBlockIndex = -1
    }
    func setSelectedBlockIndex(_ i: Int) { state.selectedBlockIndex = i }
    func setSelectedScanTypeIndex(_ i: Int) { state.selectedScanTypeIndex = i }
    func setRowWidth(_ v: String) { state.rowWidth = v }
    func setRowHeight(_ v: String) { state.rowHeight = v }
    func setBayLength(_ v: String) { state.bayLength = v }
    func setAddLens(_ v: Bool) {
        state.addLens = v
        if !v { state.multiplier = "" }
    }
    func setMultiplier(_ v: String) { state.multiplier = v }

    // MARK: - Submit

    func submit() {
        guard let data = state.data else {
            state.error = "Missing cached site data."
            return
        }
        let sites = data.sites
        let scanTypes = data.scan_type
        guard sites.indices.contains(state.selectedSiteIndex) else { return }
        let blocks = sites[state.selectedSiteIndex].blocks
        guard blocks.indices.contains(state.selectedBlockIndex),
              scanTypes.indices.contains(state.selectedScanTypeIndex) else { return }
        guard let rw = Double(state.rowWidth),
              let rh = Double(state.rowHeight),
              let bl = Double(state.bayLength) else { return }
        let multiplier = state.addLens ? (Double(state.multiplier) ?? 1.0) : 1.0

        let body = SiteBlockCreateRequest(
            blockId: blocks[state.selectedBlockIndex].block_id,
            scanTypeId: scanTypes[state.selectedScanTypeIndex].id,
            rowWidth: rw,
            rowHeight: rh,
            bayLength: bl,
            addedMultiplier: multiplier
        )

        state.isLoading = true
        state.error = nil
        do {
            let json = try String(data: JSONEncoder.tau.encode(body), encoding: .utf8) ?? ""
            ApiCache.shared.saveSubmitSiteBlock(json)
            ApiCache.shared.saveUseMainRearCameraForScan(state.addLens)
            state.isLoading = false
            state.navigateNext = true
        } catch {
            state.isLoading = false
            state.error = "Failed to save settings: \(error.localizedDescription)"
        }
    }

    func consumeNavigation() { state.navigateNext = false }
}
